### PR TITLE
Feature/show drone status live

### PR DESCRIPTION
**Pull request by Dragoo417**

### CODE REVIEW

File: app/src/main/java/ch/epfl/sdp/OfflineManagerActivity.kt [hunk 1]
```diff
@@ -63,7 +63,7 @@ class OfflineManagerActivity : MapViewBaseActivity(), OnMapReadyCallback {
             listButton = findViewById(R.id.list_button)
             listButton.setOnClickListener { downloadedRegionList() }
         }
-        setupCameraAsLastTimeUsed(this, mapboxMap)
+        mapboxMap.cameraPosition =  getLastCameraState(this)
```
Comment (Ph0tonic) [inline app/src/main/java/ch/epfl/sdp/OfflineManagerActivity.kt:66]: Same as before, you do not need to pass a context.

File: app/src/main/java/ch/epfl/sdp/ui/maps/MapUtils.kt [hunk 2]
```diff
@@ -48,13 +61,13 @@ object MapUtils {
...
     /**
-     * Loads and applies the camera settings stored in the preferences to the camera of the given
-     * map
+     * Returns the last camera state stored in the preferences
      */
-    fun setupCameraAsLastTimeUsed(context: Context, mapboxMap: MapboxMap) {
-        setupCameraWithParameters(mapboxMap,
-                loadLastMapPositionFromPrefs(context),
-                loadLastMapZoomFromPrefs(context))
+    fun getLastCameraState(context: Context): CameraPosition{
```
Comment (Ph0tonic) [inline app/src/main/java/ch/epfl/sdp/ui/maps/MapUtils.kt:66]: Same do not need to get context by parameter.

File: app/src/main/java/ch/epfl/sdp/ui/maps/MapUtils.kt [hunk 3]
```diff
@@ -69,4 +82,51 @@ object MapUtils {
                 .zoom(zoom)
                 .build()
     }
+
+    /**
+     * Creates and adds the layers necessary to display the heatmap information of the signal
+     */
+    //TODO replace context with call to MainApplication.getContext()
+    fun createLayersForHeatMap(style: Style, context: Context) {
```
Comment (Ph0tonic) [inline app/src/main/java/ch/epfl/sdp/ui/maps/MapUtils.kt:90]: Same.

File: app/src/androidTest/java/ch/epfl/sdp/OfflineManagerActivityTest.kt [hunk 4]
```diff
@@ -27,99 +26,96 @@ import org.junit.runners.MethodSorters
...
 
     @get:Rule
-    var mActivityRule = ActivityTestRule(
-            OfflineManagerActivity::class.java,
-            true,
-            true) // Activity is not launched immediately
+    var mActivityRule = IntentsTestRule(OfflineManagerActivity::class.java)
 
     private fun getContext(): Context {
         return InstrumentationRegistry.getInstrumentation().targetContext
```
[the remark this answers is not in the document]
Comment (Dragoo417) [reply app/src/androidTest/java/ch/epfl/sdp/OfflineManagerActivityTest.kt:57]: Removed

File: app/src/main/java/ch/epfl/sdp/MapActivity.kt [hunk 5]
```diff
@@ -124,17 +144,20 @@ class MapActivity : MapViewBaseActivity(), OnMapReadyCallback {
...
+            createSourceData(style)
 
             /**THIS IS JUST TO ADD SOME POINTS, IT WILL BE REMOVED AFTERWARDS**/
             addPointToHeatMap(8.543434, 47.398979)
             addPointToHeatMap(8.543934, 47.398279)
             addPointToHeatMap(8.544867, 47.397426)
             addPointToHeatMap(8.543067, 47.397026)
+
+
+            MapUtils.createLayersForHeatMap(style, this)
```
[the remark this answers is not in the document]
Comment (Dragoo417) [reply app/src/main/java/ch/epfl/sdp/MapActivity.kt:156]: Replaced this with call to Mainactivity

File: app/src/main/java/ch/epfl/sdp/OfflineManagerActivity.kt (hunk 1 above)
Comment (Dragoo417) [reply app/src/main/java/ch/epfl/sdp/OfflineManagerActivity.kt:66]: Fixed

File: app/src/main/java/ch/epfl/sdp/ui/maps/MapUtils.kt (hunk 2 above)
Comment (Dragoo417) [reply app/src/main/java/ch/epfl/sdp/ui/maps/MapUtils.kt:66]: Fixed

File: app/src/main/java/ch/epfl/sdp/ui/maps/MapUtils.kt (hunk 3 above)
Comment (Dragoo417) [reply app/src/main/java/ch/epfl/sdp/ui/maps/MapUtils.kt:90]: Done

File: app/src/main/java/ch/epfl/sdp/MapActivity.kt [hunk 6]
```diff
@@ -144,6 +167,10 @@ class MapActivity : MapViewBaseActivity(), OnMapReadyCallback {
         findViewById<TextView>(buttonId).text = (getString(labelId) + " " + DECIMAL_FORMAT.format(value))
```
[the remark this answers is not in the document]
Comment (Dragoo417) [reply app/src/main/java/ch/epfl/sdp/MapActivity.kt:167]: Replaced

File: app/src/main/java/ch/epfl/sdp/MapActivity.kt [hunk 7]
```diff
@@ -260,23 +270,20 @@ class MapActivity : MapViewBaseActivity(), OnMapReadyCallback {
      * @param newLatLng new position of the vehicle
      */
     private fun updateVehiclePosition(newLatLng: LatLng) {
-        if (droneCircleManager == null) {
-            // Not ready
-            return
-        }
+        if(!isMapReady) return
```
Comment (Ph0tonic) [inline app/src/main/java/ch/epfl/sdp/MapActivity.kt:273]: This should be `::droneCircleManager.isInitialized`

File: app/src/main/java/ch/epfl/sdp/MapActivity.kt [hunk 8]
```diff
@@ -289,10 +296,7 @@ class MapActivity : MapViewBaseActivity(), OnMapReadyCallback {
      * Updates the user position if the drawing managers are ready
      */
     private fun updateUserPosition(userLatLng: LatLng) {
-        if (userCircleManager == null) {
-            // Not ready
-            return
-        }
+        if(!isMapReady) return
```
Comment (Ph0tonic) [inline app/src/main/java/ch/epfl/sdp/MapActivity.kt:299]: Same here, `::userCircleManager.isInitialized`

File: app/src/main/java/ch/epfl/sdp/MapActivity.kt [hunk 9]
```diff
@@ -188,6 +190,8 @@ class MapActivity : MapViewBaseActivity(), OnMapReadyCallback {
      * Draws the path given by the list of positions
      */
     private fun drawPath(path: List<LatLng>) {
+        if(!isMapReady) return
```
Comment (Ph0tonic) [inline app/src/main/java/ch/epfl/sdp/MapActivity.kt:193]: Same here, `::lineManager.isInitialized`

File: app/src/main/java/ch/epfl/sdp/MapActivity.kt [hunk 10]
```diff
@@ -111,6 +112,7 @@ class MapActivity : MapViewBaseActivity(), OnMapReadyCallback {
         userLongitudeTextView = findViewById(R.id.tv_longitude)
 
         mapView.contentDescription = MAP_NOT_READY_DESCRIPTION
+        isMapReady = true
```
Comment (Ph0tonic) [inline app/src/main/java/ch/epfl/sdp/MapActivity.kt:115]: This shouldn't be here but instead in `onMapReady`

File: app/src/main/java/ch/epfl/sdp/MapActivity.kt [hunk 11]
```diff
@@ -197,6 +201,8 @@ class MapActivity : MapViewBaseActivity(), OnMapReadyCallback {
      * Fills the regions described by the list of positions
      */
     private fun drawRegion(corners: List<LatLng>) {
+        if(!isMapReady) return
+
         val fillOption = FillOptions()
                 .withLatLngs(listOf(waypoints))
                 .withFillColor(ColorUtils.colorToRgbaString(Color.WHITE))
```
Comment (Ph0tonic) [inline app/src/main/java/ch/epfl/sdp/MapActivity.kt:208]: Here are some remaining `?.` operators.

File: app/src/main/java/ch/epfl/sdp/MapActivity.kt [hunk 12]
```diff
@@ -222,16 +228,20 @@ class MapActivity : MapViewBaseActivity(), OnMapReadyCallback {
...
 
     /**
      * Clears the waypoints list and removes all the lines and points related to waypoints
      */
     private fun clearWaypoints() {
+        if(!isMapReady) return
+        
         waypoints.clear()
         waypointCircleManager?.deleteAll()
         lineManager?.deleteAll()
```
Comment (Ph0tonic) [inline app/src/main/java/ch/epfl/sdp/MapActivity.kt:247]: Can remove `?.`

**Pull request merged**